### ISSUE TITLE
Compatibility with the Array API standard?

### ISSUE THREAD

**Issue by kandersolar** [pvlib/pvlib-python#2515]
The [Array API standard](https://data-apis.org/array-api/latest/) makes it possible to write generic code that works with many types of arrays (NumPy, Tensorflow, PyTorch, Dask, JAX, CuPy, MXNet, Xarray, ...).  In principle, this would enable a pvlib simulation to be run in parallel, or on a GPU, using only code that looks very similar to the numpy-style code we are familiar with.

The Array API standard is gaining some meaningful traction in the community.  For example, SciPy is adding support: see their [array types tag](https://github.com/scipy/scipy/issues?q=label%3A%22array%20types%22), [progress tracking tables](https://scipy.github.io/devdocs/dev/api-dev/array_api.html), and [developer documentation](https://docs.scipy.org/doc/scipy/dev/api-dev/array_api.html).  scikit-learn is [working on it too](https://github.com/scikit-learn/scikit-learn/issues?q=label%3A%22Array%20API%22).

Adopting the standard could be a fairly straightforward way of enabling new use cases for pvlib.  However, I don't have any experience with it myself and can't say anything more concrete than it seems worth looking into.  So, following discussion with @echedey-ls, a good first step here is to explore the cost and benefit by prototyping compatibility in a small number of representative pvlib functions (I leave it an open question which functions might be considered representative).  That would inform decisions about whether and how to pursue broader adoption across pvlib.

For reference, I copy this description from our GSoC 2025 [project ideas list](https://github.com/pvlib/pvlib-python/wiki/GSoC-2025-Projects#project-title-compatibility-with-the-array-api):

> **Project Title: Compatibility with the Array API**
>
> **Project Description**: An exciting development in the Python ecosystem is the [Python array API standard](https://data-apis.org/array-api/latest/), a uniform interface across the various array libraries (NumPy, Xarray, JAX, etc). Historically, pvlib has been written for compatibility with NumPy arrays and Pandas Series, but making use of the Array API to expand compatibility to other array libraries could be very beneficial. A successful proposal for this project will include some investigation about what it would take to adopt the Array API across pvlib.
>
> **Expected Outcomes**: A plan/roadmap for adopting the Array API, Array API compatibility in several pvlib modules, and automated tests ensuring compatibility across the various array packages.

**Comment by kandersolar** [comment]
To the community: does this idea sound like it might be useful to your work?  Please chime in!

**Comment by AdamRJensen** [comment]
> To the community: does this idea sound like it might be useful to your work?  Please chime in!

Yes very much so. I'd like to use polars more.

**Comment by kandersolar** [comment]
Unfortunately polars is not a part of this discussion.  Unless I am mistaken, it is one of the few array libraries that does not support the Array API standard.

**Comment by cwhanse** [comment]
I'm in favor because I think that adopting the Array API standard makes pvlib functions more suitable for automatic differentiation.

**Comment by adriesse** [comment]
Sounds promising!  Beyond pandas, compatibility with xarray and plain numpy are useful to me currently.

**Comment by echedey-ls** [comment]
Preliminary research for potential functions to test on; I've got all the numpy functions and usages per function everywhere in `pvlib` except for `pvlib.iotools`, that also includes private functions (that makes it harder to assign each usage to a public function, so I haven't tried solving that).

As an insight on the current status, this is an histogram of the number of functions (y) that have a number of unique, distinct numpy functions used in their source code (x):

<img width="727" height="409" alt="Image" src="https://github.com/user-attachments/assets/ddff06ed-a2a5-42e7-bffc-df0ee4ce8285" />

The 14-different-numpy-methods function is `pvlib.iam.marion_integrate`.
There are many others in the top:
- `pvlib.pvarray.curve_fit`, 12
- `pvlib.shading.masking_angle_passias`, `pvlib.singlediode._lambertw_v_from_i`, `pvlib.solarposition.ephemeris`, 11
- `pvlib.pvsystem.sapm`, 10

I'd choose `pvlib.iam.marion_integrate`. It has `with np.errstate(invalid='ignore')` and _let's output a `pd.Series` if the input was so_ patterns. I find them relevant enough to review for this case.

On the other side there are a hella lot of em. I'd just go with a 3-numpies one, e.g. `pvlib.atmosphere.windspeed_powerlaw`.

1. [functions_analysis.py](https://github.com/user-attachments/files/21846181/functions_analysis.py) - code
2. [numpy_usage_analysis.csv](https://github.com/user-attachments/files/21846179/numpy_usage_analysis.csv) - plain results
3. [numpy_usage_analysis.ods](https://github.com/user-attachments/files/21846327/numpy_usage_analysis.ods) - results with steroids, graph and colors.

**Comment by adriesse** [comment]
I did a _little_ bit of reading, but it isn't very clear to me yet.  

A compatible pvlib function would have to use the subset of common generic functions that is defined and available through the array API.  Special strengths of individual back-end libraries could not be used. Also, pandas doesn't seem to be on the list of packages supporting the API, so that would require some extra thought.

It _seems_ like supporting the API would require rebuilding pvlib from the bottom up, although you may not have to go all the way up and could just provide partial support.

Did I get any of that right?

**Comment by echedey-ls** [comment]
@adriesse you're on the right track.

>  Special strengths of individual back-end libraries could not be used.

Yes, unique and non-API-conforming algorithms provided by each back-end lib can't be used. I guess pvlib mostly relies on common operations thou.

> pandas doesn't seem to be on the list of packages supporting the API, so that would require some extra thought.

Bingo. Kevin told me about [`pandas.DataFrame.rolling`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rolling.html), which seems a convoluted way to do a convolution. It has a nice feature thou, that it can use timedeltas as an input (in contrast to just the index-width of the window). There is also one use to calculate the wavelet function. I've never heard of it before, so IDK how feasible it is to Array-API it.

<details><summary><code>pandas.DataFrame.rolling</code> usages (4)</summary>
<p>

```bash
pvlib-python> git grep -n \.rolling
pvlib/scaling.py:280:        df = cs_long.rolling(window=intvlen, center=True, min_periods=1).mean()
pvlib/soiling.py:71:    accum_rain = rainfall.rolling(rain_accum_period, closed='right').sum()
pvlib/soiling.py:185:    accumulated_rainfall = rainfall.rolling(
pvlib/soiling.py:198:    grace_windows = rain_events.rolling(grace_period, closed='right').sum() > 0
```

</p>
</details>

**Comment by kunalKumar-13** [comment]
@echedey-ls counted pvlib's numpy usage above, and @adriesse raised the question that follows from it — *which* of those functions the Array API actually provides. I cross-referenced the two, so here is the gap with numbers on it.

Method: walk the AST of every module under `pvlib/` except `iotools`, collect each `np.<attr>`, and check each name against `array_api_strict` 2.6.1, which is the standard's reference implementation. **118 distinct attributes, 1458 usages.**

| | usages | share | distinct |
|---|---|---|---|
| already in the standard | 993 | **68.1%** | 66 |
| rename only | 178 | 12.2% | 19 |
| expressible in standard ops | 157 | 10.8% | 16 |
| **needs a decision** | **130** | **8.9%** | **17** |

**About 91% is mechanical.** That is a more encouraging picture than I expected going in.

**Rename only** — the standard has these under C-style names, so the edit is a find-and-replace once a namespace is in hand: `arctan2→atan2`, `arccos→acos`, `arcsin→asin`, `arctan→atan`, `absolute→abs`, `power→pow`, `around→round`, `mod→remainder`, `concatenate→concat`, `cumsum→cumulative_sum`, `transpose→permute_dims`, `array/asanyarray→asarray`, `dot→matmul`/`vecdot`, `vstack`/`column_stack→stack`, `append→concat`, `atleast_2d→expand_dims`, `unique→unique_values`.

**Expressible** — no direct equivalent but nothing to design. The largest by far is the degree/radian family (`radians` 66, `degrees` 26, `rad2deg` 5, `deg2rad` 2 — 99 usages) which is a multiplication by `pi/180`. Also `atleast_1d` 31, `isclose`/`allclose`, `fmax`/`fmin`, `pad`, `array_split`, `nan_to_num`, and `nanmax`/`nansum` (3 usages, and the NaN semantics differ, so they are not quite free).

**The 9% that needs a decision**, in usage order:

| | uses | why it is awkward |
|---|---|---|
| `np.isscalar` | 47 | not an array operation at all — it is a host-language type check. The standard has no notion of it |
| `np.errstate` | 34 | no equivalent. Floating-point error state is outside the standard's scope |
| `np.interp` | 12 | not in the standard |
| `np.linalg` | 10 | see below |
| `np.isreal`, `np.iscomplex` | 6 | closest is `isdtype`, but the semantics differ — these inspect values, not dtypes |
| `np.polynomial`, `polyval`, `polyfit` | 8 | no polynomial support in the standard |
| `np.vectorize`, `np.piecewise`, `np.digitize`, `np.cov` | 7 | no equivalents |
| `np.datetime64`, `np.timedelta64` | 3 | the standard has no datetime dtypes |
| `np.ndarray`, `np.dtype` | 3 | used for isinstance checks; `Array` and `isdtype` are the analogues |

Two things stand out.

**The hard part is not the maths.** `isscalar` and `errstate` are 81 of those 130 usages — 62% — and neither is an array operation. Scalar handling and floating-point error state are language- and environment-level concerns that the standard deliberately does not cover. Whatever shape adoption takes, most of the genuinely new design is in those two, not in the numerics.

**`linalg` is nearly fine, except for the one function pvlib leans on.** The standard's `linalg` extension provides `eig`, `eigh`, `svd`, `qr`, `cholesky`, `solve`, `pinv`, `inv`, `matrix_rank` — but **not `lstsq`**. pvlib's usage is `np.linalg.lstsq` × 9 and `np.linalg.eig` × 1, all in `ivtools/sde.py`, `ivtools/sdm/pvsyst.py` and `_fit_desoto_pvsyst_sandia.py`. So the covered function is the one used once and the uncovered one is used nine times. `lstsq` can be written via `pinv` or normal equations with `solve`, but not without changing conditioning, which is a real decision rather than a rewrite.

On @adriesse's point about pandas: that is the other half, and this count does not speak to it — these are numpy call sites only.

Happy to publish the script, or to re-run it including `iotools` or broken down per module if that would be more useful for scoping. It is about forty lines of `ast` plus a set membership test, so it is easy to re-run as things change.

*Disclosure, per the AI notice on the GSoC ideas page: I used AI assistance for this. The numbers come from running the analysis, not from a model's recollection, and the `lstsq` and rename findings are checked against `array_api_strict` directly — but please treat the classification boundaries as my judgement calls and push back where they look wrong.*
